default/AI/ShipDesignAI.py

```diff
@@ -1308,6 +1256,28 @@ def _calc_rating_for_name(self):
         self.update_stats(ignore_species=True)
         return self.structure
 
+    def _adjusted_production_cost(self):
+        """Return a production cost adjusted by the number of ships we have.
+
+        Building one warship of rating 2R and cost 2C is effectively cheaper than building 2 warships of rating R and
+        cost C due to fleet upkeep considerations even if they have the same raw rating/cost ratio.
+        The significance of this fleet upkeep efficiency increases with the raw fleet upkeep rate, with the expected
+        lifetime of the ship under consideration (i.e., with a longer expected lifespan it will affect the future
+        construction of a greater number of ships), and it also increases with the number of ships of this design
+        expected to be created. In the calculation below, the empire's current total shipcount is taken as a rough proxy
+        for both of the duration and extent factors.
+        Note: This same sort of adjustment would be valid for troop ships, could theoretically have some applicability
+        to scout ships since there could conceivably be a tradeoff for more scout ships of lower rating/range, but
+        would really not be applicable to colony ships since in that case there is really not a potential tradeoff
+        between number of ships and capacity.
+
+        :return: adjusted production cost
+        :rtype: float
+        """
+        # TODO: Consider total pp production output as additional factor
```

Comment:
good point about totap pp production  -- we might want to rule out designs that would consumer the empire's entire production for too many turns in a row.  That example is rather extreme but the cases can be more subtle as well-- the higher the per-turn cost of the ship relative to the empires total pp, then there is less flexibility to work other construction projects into the production queue, which somewhat increases the effective opportunity cost of producing the ship.  It's not clear to me though if we would really want to handle that latter issue here in the ShipDesigner code though-- so long as two  otherwise similarly rated designs were presented to the production choice code, that might be the best place assess issues re flexibility of production.  Putting something here to potentially increase the adjusted cost is probably the simplest way to start addressing the issue, though, even if the handling did eventually move.